plugins/optimization-detective/class-od-url-metric.php

```diff
@@ -44,45 +44,51 @@
  *                                elements: ElementData[]
  *                            }
  *
+ * @property-read string        $uuid
+ * @property-read string        $url
+ * @property-read float         $timestamp
+ * @property-read ViewportRect  $viewport
+ * @property-read ElementData[] $elements
```

Comment:
Not sure about the value of this and `__get()`. Seems unrelated to the purpose of this PR anyway, so maybe let's discuss that separately? I think the `get()` method is good enough?

Reply:
In https://github.com/WordPress/performance/pull/1098#discussion_r1550492139 you suggested `__call()` and I responded there.

Reply:
That's fair. This works.

My only concern with such an approach is that it only works reasonably well if another plugin extending it would also have its own class that it annotated with these `@property-read` annotations. Otherwise, it becomes an incomplete documentation, and in that case I feel it might even be better to just have `get()`, so that it's more obvious to a developer they need to "manually" look at what's available - whereas these annotations may imply that you can simply rely on them, when as soon as an extension happens, this may not longer be the case.

So I think I still prefer not to have this, but not a big deal if you want to keep it.

Reply:
That's true. Plugins that extend the schema wouldn't have type definitions for these properties. However, having them for the core properties is useful so that we don't have to add `get_*()` methods for every root property. I just added 384fa53f9de60b3080b515ab95a06070b50b1a04 & ed5815158552d27359a730e1f6dfa41ccbcc61c9 to make use of these. So it's more ergonomic and feels better. (I think we can deprecate `get_uuid()`, `get_timestamp()`, `get_elements()`, etc.) 

For plugins that extend the schema, this ties back to the discussion of having default values for the properties. This makes me think that we should force the default to be `null` so that `isset()` checks will be `false`. It doesn't make sense to have a non-`null` default defined when the client never submitted data for that URL metric in the first place. Having a default value of `null` will avoid cases where array indices will be undefined, but at the same time `isset()` checks will be `false`.

As for plugins that extend the schema, I'm thinking about Embed Optimizer adding a `resizedBoundingClientRect` property for the element. This isn't actually directly connected to the `@property-read` discussion, but it is related in static analysis. The Embed Optimizer plugin could access it from an element via:

```php
foreach ( $url_metric->elements as $element ) {
	if ( null !== $element['resizedBoundingClientRect'] ) {
		// ...
	}
}
```

However, PHPStan here complains:

> phpstan: Offset 'resizedBoundingClie…' does not exist on array{isLCP: bool, isLCPCandidate: bool, xpath: string, intersectionRatio: float, intersectionRect: array{width: float, height: float, x: float, y: float, top: float, right: float, bottom: float, left: float}, boundingClientRect: array{width: float, height: float, x: float, y: float, top: float, right: float, bottom: float, left: float}}.

This makes sense because `$element` is typed as `ElementData` which does not have `resizedBoundingClientRect`. (Aside: I think `ElementData` should perhapsbe renamed to  `ElementMetrics`.) 

I tried adding the following to the `Embed_Optimizer_Tag_Visitor` class:

```
 * @phpstan-import-type DOMRect from OD_URL_Metric
 * @phpstan-import-type ElementData from OD_URL_Metric
 * @phpstan-type ExtendedElementData ElementData & array{ resizedBoundingClientRect: ?DOMRect }
```

And then explicitly overriding the type for `$element`:

```php
/**
 * Extended element.
 *
 * @var ExtendedElementData $element
 */
```

But it's not working. I'm getting either errors like:

```
PHPDoc tag @var for variable $element contains unresolvable type.
```

Or when moving `@phpstan-type ExtendedElementData` to the method, I instead get:

```
phpstan: Access to offset 'resizedBoundingClie…' on an unknown class ExtendedElementData.
phpstan: PHPDoc tag @var for variable $element contains unknown class ExtendedElementData.
phpstan: PHPDoc tag @var with type ExtendedElementData is not subtype of type array{isLCP: bool, isLCPCandidate: bool, xpath: string, intersectionRatio: float, intersectionRect: array{width: float, height: float, x: float, y: float, top: float, right: float, bottom: float, left: float}, boundingClientRect: array{width: float, height: float, x: float, y: float, top: float, right: float, bottom: float, left: float}}.
```

I'm nearing my PHPStan knowledge limits. Maybe it just can't do what I want.

Reply:
I think it comes down to the aforementioned problem with magic getters and `@property-read`. You can't programmatically guarantee that any of this will work. Technically every single use of the magic getter would need to be wrapped in a magic "isset-er".

That's related to my previous concern, and I don't think we should deprecate the getter methods in favor of magic properties. I think the win in ergonomics is tiny - only about not having to write `get_`, which IMO is not a good reason to change this, given that using the methods is a more reliable approach, which PHPStan should be fine with.

For any custom properties, we can use the `get( ... )` method, which is not ideal either, but still could facilitate a cleaner approach where an extension would implement its own class that implements additional `get_...()` methods for new properties.

Reply:
Actually I think the problem here is not related to `@property-read`, since this is regarding accessing keys of the items in the `elements` array. It's the same problem if using `get_elements()` with this code:

```php
foreach ( $url_metric->get_elements() as $element ) {
	if ( null !== $element['resizedBoundingClientRect'] ) {
		// ...
	}
}
```

The problem is that `$element` is typed as `ElementData` which doesn't include the `resizedBoundingClientRect` key. But this code here is assuming the `null` was provided as a `default`. Let's consider that it may not be set at all. In that case, consider this code:

```php
foreach ( $url_metric->elements as $element ) {
	if ( isset( $element['resizedBoundingClientRect'] ) ) {
		/**
		 * Resized bounding client rect.
		 *
		 * @var DOMRect $rect
		 */
		$rect = $element['resizedBoundingClientRect'];
		PHPStan\dumpType( $rect );
	}
}
```

Here we have the `isset()` check and considering that the schema has enforced the expected type for `resizedBoundingClientRect`, PHPStan here outputs:

```
Dumped type: array{width: float, height: float, x: float, y: float, top: float, right: float, bottom: float, left: float}
```

This is what we want. However, it would be ideal if we didn't have to use `@var` and PHPStan could be told directly that `$element['resizedBoundingClientRect']` is of type `DOMRect|null`.

Otherwise, for built-in root-level keys like `elements`, `uuid`, `timestamp` they will never return `null` because the schema enforces them to have a value. For the non-builtin properties, however, they indeed may not be set. So let's say there's a new root property called `cwv_metrics` which the Dashboard (https://github.com/WordPress/performance/issues/1324) may introduce. This could be accessed in this PR in the following ways:

```php
PHPStan\dumpType( $url_metric->get( 'web_vitals' ) );
```

```
Dumped type: mixed
```

No type information as expected since `get()` returns `mixed|null`. However, this with the magic getter is worse, even though `__get()` is defined the same:

```php
PHPStan\dumpType( $url_metric->web_vitals )
```

```
Access to an undefined property OD_URL_Metric::$web_vitals.
💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property  
Dumped type: *ERROR*
```

That helpful [Learn more](https://phpstan.org/blog/solving-phpstan-access-to-undefined-property) link may have the answer and I'll investigate further tomorrow. In particular, it may come down to [writing our own PHPStan extension](https://phpstan.org/blog/solving-phpstan-access-to-undefined-property#write-your-own-extension). This could ensure that PHPStan detects the correct types for both `get()` and magic getters, as well as for extended `elements` item schemas.

Reply:
Writing a PHPStan extension sounds like overkill to me. I think we should try to make it work against its constraints.

Regarding the `mixed` return value of  the `get()` method, I don't think that's a problem. Wouldn't an extension still be able to introduce a wrapper methods where it casts it to what's expected and comes with proper return type hint and doc? I think that's sufficient to alleviate the concerns.

And for extension that don't want to go the extra mile, they can still do it the "dirty" way of directly calling `get()` wherever they need an extra property and then checking for existence and cast the value as needed.

On the other hand, undefined properties is a real problem, as reported by PHPStan. That's why I would continue to advise against using a magic getter. I just don't see the value.

Reply:
The magic getter is only relevant to the root properties. The issue exists for extensions to `element` data. An extension plugin could indeed do its own type declarations in helper wrapper functions, for example:

<!--

```php
	/**
	 * Gets resized bounding client rect from an element.
	 *
	 * @phpstan-param ElementData $element
	 *
	 * @param array $element Element.
	 * @return DOMRect
	 */
	public static function get_resized_bounding_client_rect( array $element ): ?array {
		if ( isset( $element['resizedBoundingClientRect'] ) ) {
			/**
			 * Resized bounding client rect.
			 *
			 * @var DOMRect $rect
			 */
			$rect = $element['resizedBoundingClientRect'];
			return $rect;
		}
		return null;
	}

	/**
	 * Gets minimum resized height.
	 *
	 * @param OD_URL_Metric $url_metric URL Metric.
	 * @return float|null Minimum resized height or null if no resized information was captured.
	 */
	public function get_minimum_resized_height( OD_URL_Metric $url_metric ): ?float {
		$min_resized_height = null;
		foreach ( $url_metric->get_elements() as $element ) {
			$rect = self::get_resized_bounding_client_rect( $element );
			if ( isset( $rect ) ) {
				$min_resized_height = isset( $min_resized_height ) 
					? min( $min_resized_height, $rect['height'] ) 
					: $rect['height'];
			}
		}
		return $min_resized_height;
	}
```

But it would be nicer if we could eliminate the need for `get_resized_bounding_client_rect()` and just do:

```php
	/**
	 * Gets minimum resized height.
	 *
	 * @param OD_URL_Metric $url_metric URL Metric.
	 * @return float|null Minimum resized height or null if no resized information was captured.
	 */
	public function get_minimum_resized_height( OD_URL_Metric $url_metric ): ?float {
		$min_resized_height = null;
		foreach ( $url_metric->get_elements() as $element ) {
			$rect = self::get_resized_bounding_client_rect( $element );
			if ( isset(  ) ) {
				$min_resized_height = min( $min_resized_height, $rect['height'] );
			}
		}
		return $min_resized_height;
	}
```

-->


```php
	/**
	 * Gets the resized bounding client rect.
	 *
	 * @phpstan-param ElementData $element
	 *
	 * @param array $element Element.
	 * @return DOMRect|null Whether visible or null if unknown.
	 */
	public static function get_resized_bounding_client_rect( array $element ): ?array {
		if ( isset( $element['resizedBoundingClientRect'] ) ) {
			return $element['resizedBoundingClientRect'];
		}
		return null;
	}

	public function do_something( OD_HTML_Tag_Processor $processor, OD_URL_Metric $url_metric ): void {
		foreach ( $url_metric->get_elements() as $element ) {
			$rect = self::get_resized_bounding_client_rect( $element );
			if ( isset( $rect ) ) {
				PHPStan\dumpType( $rect['height'] ); // => Dumped type: float
			}
		}
	}
```

It would be nicer to not have to resort to `get_resized_bounding_client_rect` and just do:

```php
	public function do_something( OD_HTML_Tag_Processor $processor, OD_URL_Metric $url_metric ): void {
		foreach ( $url_metric->get_elements() as $element ) {
			if ( isset( $element['resizedBoundingClientRect'] ) ) {
				PHPStan\dumpType( $element['resizedBoundingClientRect']['height'] );
			}
		}
	}
```

But this results in `Dumped type: mixed`.

Informing PHPStan how to resolve the extended schema would ensure the expected types are used without having to duplicate typing information which is already being enforced by validation and sanitization with the schema.

For root properties it also happens, without magic getters:

```php
	/**
	 * Gets the CWV metrics.
	 *
	 * @param OD_URL_Metric $url_metric URL Metric.
	 * @return array{ lcp: float, inp: ?float, cls: float }|null CWV data or null if not set.
	 */
	public static function get_cwv_metrics( OD_URL_Metric $url_metric ): ?array {
		$cwv_metrics = $url_metric->get( 'cwv_metrics' );
		if ( is_array( $cwv_metrics ) ) {
			/**
			 * CWV metrics.
			 *
			 * @var array{ lcp: float, inp: ?float, cls: float } $cwv_metrics
			 */
			return $cwv_metrics;
		}
		return null;
	}

	public function do_something( OD_URL_Metric $url_metric ): void {
		$cwv_metrics = self::get_cwv_metrics( $url_metric );
		if ( isset( $cwv_metrics ) ) {
			PHPStan\dumpType( $cwv_metrics['lcp'] ); // => Dumped type: float
		}
	}
```

It would be preferable to just have to do:

```php
	public function do_something( OD_URL_Metric $url_metric ): void {
		$cwv_metrics = $url_metric->get( 'cwv_metrics' );
		if ( isset( $cwv_metrics ) ) {
			PHPStan\dumpType( $cwv_metrics['lcp'] );
		}
	}
```

But again, this results in `Dumped type: mixed`.

I don't think the use of a magic getter is necessarily a problem. I believe static analysis would be able to deal with them. But it isn't necessary. I've removed it in 85a03a96. I would still like to explore the use of a PHPStan extension. It's designed to be extensible for cases like this.